R/shared-methods.R

```diff
@@ -97,9 +97,8 @@ initialise_dataclass <- function(self, level = "1", regions,
   }
 
   if (!is.null(self$level_data_urls[[self$level]])) {
-    self$data_urls <- merge(
-      self$common_data_urls,
-      self$level_data_urls[[self$level]]
+    self$data_urls <- unlist(
+      list(self$common_data_urls, self$level_data_urls[[self$level]])
     )
```

Comment:
This may become moot since I don't think South Africa needs level-specific url lists, but instead of dropping merge could we just put a `length()!=0` test along side the `is.null()` test and so still get the possible deconfliction benefits of using `merge`?

The logic I'd written at one point is below (then we shifted to nested lists but the logic flows)
```
      data_url_list <- self$data_url

      level_specific_data_url <- switch(self$level,
                                        "1" = self$data_url_level_1,
                                        "2" = self$data_url_level_2)

      if (!is.null(level_specific_data_url)) {
        if (length(self$data_url) > 0) {
          data_url_list <- merge(self$data_url, level_specific_data_url)
        } else {
          data_url_list <- level_specific_data_url
        }
      }
```

Reply:
My only concern with this approach is it adds more conditional checks to do what unlist does.

Reply:
Ah. What we want may be in `purrr::list_modify`

I think we would want to do something like `unlist(list_modify(level_specific_data_url, self$data_url))` but `list_modify` is doing something odd to the labels.

On a more general question - perhaps move to discussion - what is wrong with another conditional check here? This is code which gets run once, it's not in a loop, we're pretty far from the metal to be worrying about breaking pipelining - and I imagine that the functions we're calling are pretty thick with conditionals. I guess that `length` is probably *O(1)* (but I don't know); if it turns out to be zero we get a single copy instead of the merge and copy.

Reply:
So the thing I am keen to avoid is writing level into any of the code as this makes it harder to add more levels down the line. Other checks (like if `common_data_urls` and `level_data_urls` are present and longer than 0) are absolutely not a problem and are I think what we should be using.

Reply:
I've run into this list of list updating thing before and last time I wrote this little utility: 

https://github.com/epiforecasts/EpiNow2/blob/bee6bc765f754508da1235466a3388b16a0af812/R/utilities.R#L324

It feels like there must be a nicer way to do this and perhaps `list_modify` is it but I am very confused about the names it is assigning.

Reply:
Happy to use whatever, just used unlist as it worked at the time but there are probably better options. 

This seems to affect USA, South Africa, France and Belgium.


Should I go with richards approach in #250 and change to that?